File: 0141-linked-list-cycle/0141-linked-list-cycle.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

require 'set'
# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
    traverse = Set[];
    while !(head.equal? nil)
        if traverse.include?(head) 
            return true
        end
        traverse.add head
        head = head.next
    end
    false
end

File: 0141-linked-list-cycle/0141-linked-list-cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode ListNode;

bool hasCycle(ListNode *head) {
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast != NULL) {
        if (fast->next == NULL) {
            return false;
        }
        fast = fast->next->next;
        slow = slow->next;
        if (fast == slow) {
            return true;
        }
    }
    return false;
}